analyzers/src/SonarAnalyzer.CSharp/Rules/ExistsInsteadOfAny.cs

```diff
@@ -0,0 +1,48 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ExistsInsteadOfAny : ExistsInsteadOfAnyBase<SyntaxKind, InvocationExpressionSyntax>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+    protected override bool HasOneArgument(InvocationExpressionSyntax node) =>
+        node.HasExactlyNArguments(1);
+
+    protected override bool IsValueEquality(InvocationExpressionSyntax node, SemanticModel model) =>
+        node.ArgumentList.Arguments[0].Expression is SimpleLambdaExpressionSyntax lambda
+        && lambda.Body switch
+        {
+            BinaryExpressionSyntax binary =>
+                binary.OperatorToken.IsKind(SyntaxKind.EqualsEqualsToken)
+                && IsIdentifierOrLiteralValueType(binary.Left, model)
+                && IsIdentifierOrLiteralValueType(binary.Right, model),
+            InvocationExpressionSyntax invocation =>
+                Language.GetName(invocation).Equals(nameof(Equals), Language.NameComparison),
```

Comment:
I think this is getting better, but it still needs some improvements to be more precise. 
As far as this rule is concerned, precision is not that important, but we need to get this to raise _exactly_ on the `Contains` branch, and then throw everything else on the `Exists` branch.

Intuition: We need to cover these using `Exists`, meaning this method should return false:
```csharp
x => x == x // you marked this as a FN, but this means it will raise `Contains`
x => y == z 
x => y == 42
x => 42 == 42
```

Right now, you check distinctly if left is Literal/Identifier and then the same for the right, but it should be done together with a disjunction, to avoid these cases.

I think the algorithm should be something like this:
- Establish that you have either binary or invocation (already done)

- Get the left/right of the expression and check if they are correct, via `HasValidOperands(first, second)` (placeholder name)

This method should check the following:
`first` is `IdentifierSyntax` with name equal to the lambda's parameter (this might need symbolic-equality, but I think it's probably fine with just syntax).
`second` is either `IdentifierSyntax` with name _not_ equal to lambda's parameter OR a literal. Technically you can have a lot of other cases that you want to raise `Contains`, for example: `x => x == ApplyMethod(42)` but then you would have to parse every possible expression-like syntax and filter out expressions that contain the x on the other side as well. For example, this should raise `Contains` `x => x == A(B(C(42)))`  but this should raise `Exists` `x => x == A(B(C(x+1)))`...but it's awfully hard to track it precisely, so we can get away with raising `Exists` on all the other cases.

- After you have this method, you can invoke it by passing the arguments twice, as in `(Check(left, right) || Check(right,left))`, since you want to cover all of the (expr1 == expr2) but also (expr2 == expr1).

- Notice that this check should happen on both `==` and `Equals`. After you are done, the only thing you need to check last is `(type.IsValueType || type.Is(KnownType.System_String)`, only on the `binary ==` branch.